packages/react-server-cli/src/commands/start.js

```diff
@@ -13,239 +13,172 @@ import logProductionWarnings from "../logProductionWarnings";
+				logger.error(e);
 				reject(e);
 			});
-			httpServer.listen(port, bindIp, (e) => {
+			httpServer.close((e) => {
 				if (e) {
+					logger.error('The server was not started, so it cannot be stopped.');
+					logger.error(e);
 					reject(e);
 					return;
 				}
-				logger.info(`Started HTML server over ${httpsOptions ? "HTTPS" : "HTTP"} on ${bindIp}:${port}`);
 				resolve();
 			});
-		}),
+		});
 	};
 };
 
-// given a webpack compiler and a port, compile the JavaScript code to static
-// files and start up a web server at http://host:port/ that serves the
-// static compiled JavaScript. returns an object with two properties, started and stop;
+
+// given the server routes file and a port, start a react-server server at
+// http://host:port/. returns an object with two properties, started and stop;
 // see the default function doc for explanation.
-const startStaticJsServer = (compiler, port, bindIp, longTermCaching, httpsOptions) => {
-	const server = express();
-	const httpServer = httpsOptions ? https.createServer(httpsOptions, server) : http.createServer(server);
+const startServer = (serverRoutes, options, compiler) => {
+	const {
+		port,
+		bindIp,
+		httpsOptions,
+		customMiddlewarePath,
+		hot,
+		longTermCaching,
+	} = options;
+
+	let expressServer,
+		httpServer;
+
+	if (hot) {
+		logger.notice("Enabling hot module reload with webpack-dev-server");
+		const webpackDevServer = new WebpackDevServer(compiler, {
+			noInfo: true,
+			hot: true,
+			headers: { 'Access-Control-Allow-Origin': '*' },
+			https: !!httpsOptions,
+			key: httpsOptions ? httpsOptions.key : undefined,
+			cert: httpsOptions ? httpsOptions.cert : undefined,
+			ca: httpsOptions ? httpsOptions.ca : undefined,
+
+			// contentBase: false is required in order to prevent WebpackDevServer from trying to serve static directory
+			// entries for the URL path '/'.  When WebpackDevServer does this, Express never calls react-server to attempt
+			// to handle the request.
+			contentBase: false,
+		});
+		expressServer = webpackDevServer.app;
+		httpServer = webpackDevServer;
+		// This allows us to attach our own bindings to the .on() events from http/https server and still use the
+		// WebpackDevServer .listen() and .close(), which is required for the HMR websocket to be opened.
+		httpServer.on = webpackDevServer.listeningApp.on;
+	} else {
+		// Only compile the webpack configs manually if we're not in hot mode
+		logger.notice("Compiling Webpack bundle prior to starting server");
+		compiler.run((err, stats) => {
+			handleCompilationErrors(err, stats);
+		});
+
+		expressServer = express();
+		expressServer.use('/', compression(), express.static(`__clientTemp/build`, {
+			maxage: longTermCaching ? '365d' : '0s',
+		}));
+		httpServer = httpsOptions ? https.createServer(httpsOptions, expressServer) : http.createServer(expressServer);
+	}
+
+	let middlewareSetup = (server, rsMiddleware) => {
+		server.use(compression());
```

Comment:
I don't know if that's a problem but it looks like the `compression()` middleware is used two times in non-hot mode.